plugin/pkg/admission/allowunsafedelete/admission.go

```diff
@@ -0,0 +1,138 @@
+var _ admission.InitializationValidator = &AllowUnsafeDelete{}
+var _ genericadmissioninit.WantsAuthorizer = &AllowUnsafeDelete{}
+
+// SetAuthorizer sets the authorizer.
+func (p *AllowUnsafeDelete) SetAuthorizer(authz authorizer.Authorizer) {
+	p.authz = authz
+}
+
+// ValidateInitialization ensures an authorizer is set.
+func (p *AllowUnsafeDelete) ValidateInitialization() error {
+	if p.authz == nil {
+		return fmt.Errorf("%s requires an authorizer", PluginName)
+	}
+	return nil
+}
+
+// Validate ensures that the user has permission to do 'delete-ignore-read-errors'
+// on the resource being deeted when ignoreStoreReadErrorWithClusterBreakingPotential
+// is enabled, these are the constraints:
+// a) must be a DELETE operation
+// b) ignoreStoreReadErrorWithClusterBreakingPotential is set to true
+// c) the request is a resource, no sub-resource or non-resource endpoint
+// d) the user has permission to do 'delete-ignore-read-errors' on the resource
+func (p *AllowUnsafeDelete) Validate(ctx context.Context, attr admission.Attributes, o admission.ObjectInterfaces) (err error) {
+	// if the feature is disabled, this plugin, although
+	// enabled, should not be active
+	if !utilfeature.DefaultFeatureGate.Enabled(genericfeatures.AllowUnsafeMalformedObjectDeletion) {
+		return nil
+	}
+	if attr.GetOperation() != admission.Delete || attr.GetOperationOptions() == nil {
+		return nil
+	}
+
+	options, ok := attr.GetOperationOptions().(*metav1.DeleteOptions)
+	if !ok {
+		return fmt.Errorf("expected an option of type: %T, but got: %T", &metav1.DeleteOptions{}, attr.GetOperationOptions())
+	}
+	if ignore := options.IgnoreStoreReadErrorWithClusterBreakingPotential; ignore == nil || !*ignore {
+		return nil
+	}
+
+	requestInfo, found := request.RequestInfoFrom(ctx)
+	if !found {
+		return admission.NewForbidden(attr, errors.New("no RequestInfo found in the context"))
+	}
+	if !requestInfo.IsResourceRequest || len(attr.GetSubresource()) > 0 {
+		return admission.NewForbidden(attr, errors.New("ignoreStoreReadErrorWithClusterBreakingPotential delete option is not allowed on a subresource or non-resource request"))
+	}
+
+	// if we are here, IgnoreStoreReadErrorWithClusterBreakingPotential
+	// is set to true in the delete options, the user must have permission
+	// to do 'delete-ignore-read-errors' on the given resource.
+	const verb = "delete-ignore-read-errors"
+	record := authorizer.AttributesRecord{
+		User:            attr.GetUserInfo(),
+		Verb:            verb,
+		Namespace:       attr.GetNamespace(),
+		Name:            attr.GetName(),
+		APIGroup:        attr.GetResource().Group,
+		APIVersion:      attr.GetResource().Version,
+		Resource:        attr.GetResource().Resource,
+		ResourceRequest: true,
+	}
+
+	decision, reason, err := p.authz.Authorize(ctx, record)
+	if err != nil {
+		err = fmt.Errorf("error while checking permission for %q, %w", verb, err)
+		klog.V(1).ErrorS(err, "failed to authorize")
```

Comment:
wondering, don't we have the structured logger from the context? This kind of global klog is deprecated I believe.

Reply:
sure, added context based logging